plugins/jabra-gnp/fu-jabra-gnp-device.c

```diff
@@ -0,0 +1,859 @@
+	fu_archive_firmware_set_format(FU_ARCHIVE_FIRMWARE(firmware), FU_ARCHIVE_FORMAT_ZIP);
+	fu_archive_firmware_set_compression(FU_ARCHIVE_FIRMWARE(firmware),
+					    FU_ARCHIVE_COMPRESSION_NONE);
+
+	/* unzip and get images */
+	if (!fu_firmware_parse(firmware, fw, flags, error))
+		return NULL;
+	return g_steal_pointer(&firmware);
+}
+
+static gboolean
+fu_jabra_gnp_device_probe(FuDevice *device, GError **error)
+{
+	g_autoptr(GError) error_local = NULL;
+	FuJabraGnpDevice *self = FU_JABRA_GNP_DEVICE(device);
+	self->iface_hid =
+	    _g_usb_device_get_interface_for_class(fu_usb_device_get_dev(FU_USB_DEVICE(self)),
+						  G_USB_DEVICE_CLASS_HID,
+						  &error_local);
+	if (self->iface_hid == 0xFF) {
+		g_set_error(error,
+			    FWUPD_ERROR,
+			    FWUPD_ERROR_NOT_SUPPORTED,
+			    "cannot find HID interface: %s",
+			    error_local->message);
+		return FALSE;
+	}
+	fu_usb_device_add_interface(FU_USB_DEVICE(self), self->iface_hid);
+	return TRUE;
+}
+
+static gboolean
+fu_jabra_gnp_device_setup(FuDevice *device, GError **error)
+{
+	FuJabraGnpDevice *self = FU_JABRA_GNP_DEVICE(device);
+	if (!fu_jabra_gnp_device_read_version(self, error))
+		return FALSE;
+	return TRUE;
+}
+
+static gboolean
+fu_jabra_gnp_device_get_part_data_from_file_node(XbNode *n,
+						 FuJabraGnpPartitionData *part_data,
+						 GError **error)
+{
+	gchar *crc_str = NULL;
+	gchar *part_str = NULL;
+	part_data->file_name = xb_node_get_attr(n, "name");
+	part_data->language = xb_node_query_text(n, "language", NULL);
+	part_data->version = xb_node_query_text(n, "version", NULL);
+	part_data->target = xb_node_query_text(n, "target", NULL);
+	crc_str = xb_node_query_text(n, "crc", NULL);
+	gchar *tmp = fu_strsafe(crc_str + 2, 8);
+	if (!g_ascii_string_to_unsigned(tmp, 16, 0, 0xffffffff, &(part_data->crc), error))
+		return FALSE;
+	part_str = xb_node_query_text(n, "partition", NULL);
+	if (!g_ascii_string_to_unsigned(part_str, 10, 0, 0xff, &(part_data->partition), error))
+		return FALSE;
+	return TRUE;
+}
+
+static gboolean
+fu_jabra_gnp_device_get_data_from_info(FuJabraGnpDevice *self,
+				       XbSilo *silo,
+				       gchar *requested_version,
+				       FuJabraGnpVersionData *requested_version_data,
+				       guint64 *dfu_pid,
+				       GError **error)
+{
+	g_autoptr(GPtrArray) dfu_pids = NULL;
+	g_autoptr(GPtrArray) build_vectors = NULL;
+	gchar *tmp = NULL;
+	build_vectors = xb_silo_query(silo, "buildVector", 0, error);
+	XbNode *n = g_ptr_array_index(build_vectors, 0);
+	requested_version = xb_node_get_attr(n, "version");
+	if (!fu_jabra_gnp_device_parse_version(requested_version, requested_version_data, error))
+		return FALSE;
+	if (!fu_jabra_gnp_device_read_dfu_pid(self, error))
+		return FALSE;
+	dfu_pids = xb_silo_query(silo, "buildVector/targetUsbPids", 0, error);
+	if (dfu_pids == NULL)
+		return FALSE;
+
+	for (guint i = 0; i < dfu_pids->len; i++) {
+		n = g_ptr_array_index(dfu_pids, i);
+		tmp = xb_node_query_text(n, "usbPid", NULL);
```

Comment:
is there ever more than one usbPid in the targetUsbPids block?

Reply:
As far as I am aware, there may be. I'm trying to get confirmation internally.

Reply:
Okay, lets push this with one supported, and if we need to support both we can make it more complicated.